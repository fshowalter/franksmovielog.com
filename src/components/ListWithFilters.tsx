import type { JSX, ReactNode } from "react";

import { useCallback, useEffect, useRef, useState } from "react";

export const DRAWER_CLOSE_ANIMATION_MS = 250;
const DRAWER_OPEN_ANIMATION_MS = 400;

type Props<T extends string> = {
  className?: string;
  dynamicSubNav?: React.ReactNode;
  filters: React.ReactNode;
  hasActiveFilters?: boolean;
  list: React.ReactNode;
  listHeaderButtons?: React.ReactNode;
  onApplyFilters?: () => void;
  onClearFilters?: () => void;
  onFilterDrawerOpen?: () => void;
  onResetFilters?: () => void;
  pendingFilteredCount?: number;
  sortProps: SortProps<T>;
  totalCount: number;
};

type SortProps<T extends string> = {
  currentSortValue: T;
  onSortChange: React.ChangeEventHandler<HTMLSelectElement>;
  sortOptions: React.ReactNode;
};

// This is used by pages directly, keep it exported
export function ListHeaderButton({
  href,
  text,
}: {
  href: string;
  text: string;
}) {
  return (
    <div
      className={`
        flex items-start gap-x-4 bg-default px-4 text-nowrap text-accent
        uppercase
      `}
    >
      <a
        className={`
          relative inline-block transform-gpu py-1 transition-transform
          after:absolute after:bottom-0 after:left-0 after:h-px after:w-full
          after:origin-center after:scale-x-0 after:bg-(--fg-accent)
          after:transition-transform
          hover:after:scale-x-100
        `}
        href={href}
      >
        {text}
      </a>
    </div>
  );
}

export function ListWithFilters<T extends string>({
  className,
  dynamicSubNav,
  filters,
  hasActiveFilters,
  list,
  listHeaderButtons,
  onApplyFilters,
  onClearFilters,
  onFilterDrawerOpen,
  onResetFilters,
  pendingFilteredCount,
  sortProps,
  totalCount,
}: Props<T>): JSX.Element {
  const [filterDrawerVisible, setFilterDrawerVisible] = useState(false);
  const [isClosing, setIsClosing] = useState(false);
  const [isOpening, setIsOpening] = useState(false);
  const filtersRef = useRef<HTMLDivElement | null>(null);
  const toggleButtonRef = useRef<HTMLButtonElement | null>(null);
  const timeoutRefs = useRef<Set<NodeJS.Timeout>>(new Set());
  const prevSortValueRef = useRef<T>(sortProps.currentSortValue);

  const handleCloseDrawer = useCallback(
    (shouldResetFilters = true) => {
      setIsClosing(true);
      // Start the spin animation, then close after a short delay
      const timeoutId = setTimeout(() => {
        if (typeof document !== "undefined") {
          document.body.classList.remove("overflow-hidden");
        }
        setFilterDrawerVisible(false);
        setIsClosing(false);
        if (shouldResetFilters) {
          onResetFilters?.();
        }
        timeoutRefs.current.delete(timeoutId);
      }, DRAWER_CLOSE_ANIMATION_MS);
      timeoutRefs.current.add(timeoutId);
    },
    [onResetFilters],
  );

  const onFilterClick = useCallback(
    (event: React.MouseEvent) => {
      event.preventDefault();

      if (filterDrawerVisible) {
        handleCloseDrawer();
      } else {
        setIsOpening(true);
        if (typeof document !== "undefined") {
          document.body.classList.add("overflow-hidden");
        }
        setFilterDrawerVisible(true);
        // Call onFilterDrawerOpen when opening
        onFilterDrawerOpen?.();
        // Clear the opening state after animation completes
        const timeoutId = setTimeout(() => {
          setIsOpening(false);
          timeoutRefs.current.delete(timeoutId);
        }, DRAWER_OPEN_ANIMATION_MS);
        timeoutRefs.current.add(timeoutId);
        // Focus first focusable element after drawer opens
        requestAnimationFrame(() => {
          const firstFocusable = filtersRef.current?.querySelector<HTMLElement>(
            'button:not([disabled]), input:not([disabled]), select:not([disabled]), textarea:not([disabled]), [tabindex]:not([tabindex="-1"])',
          );
          firstFocusable?.focus();
        });
      }
    },
    [filterDrawerVisible, handleCloseDrawer, onFilterDrawerOpen],
  );

  // Cleanup timeouts on unmount
  useEffect(() => {
    return () => {
      for (const timeoutId of timeoutRefs.current) clearTimeout(timeoutId);
      timeoutRefs.current.clear();
    };
  }, []);

  // Handle escape key
  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === "Escape" && filterDrawerVisible && !isClosing) {
        handleCloseDrawer();
        toggleButtonRef.current?.focus();
      }
    };

    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [filterDrawerVisible, handleCloseDrawer, isClosing]);

  // Scroll to top of list when sort changes
  useEffect(() => {
    if (prevSortValueRef.current !== sortProps.currentSortValue) {
      prevSortValueRef.current = sortProps.currentSortValue;
      if (typeof document !== "undefined") {
        document.querySelector("#list")?.scrollIntoView({ behavior: "smooth" });
      }
    }
  }, [sortProps.currentSortValue]);

  return (
    <div
      className={`
        ${className || ""}
      `}
    >
      {dynamicSubNav}
      <div className={`group/list-with-filters mx-auto bg-subtle`}>
        <div
          className={`
            sticky top-[calc(0px_+_var(--scroll-offset,0px))] z-sticky
            scroll-mt-[calc(0px_+_var(--scroll-offset,0px))] border-b
            border-default bg-default text-xs
            tablet:col-span-full
          `}
        >
          <ListHeader
            filterDrawerVisible={filterDrawerVisible}
            listHeaderButtons={listHeaderButtons}
            onFilterClick={onFilterClick}
            sortProps={sortProps}
            toggleButtonRef={toggleButtonRef}
            totalCount={totalCount}
          />
        </div>
        <div
          className={`
            mx-auto max-w-[var(--breakpoint-desktop)]
            tablet:px-container
          `}
        >
          <div
            className={`
              mx-auto max-w-[var(--breakpoint-desktop)] grow
              scroll-mt-[calc(var(--list-scroll-offset)_+_var(--scroll-offset,0px))]
              pb-10
              [--list-scroll-offset:181px]
              tablet:[--list-scroll-offset:121px]
            `}
            id="list"
          >
            {list}
          </div>

          {/* Backdrop for filters */}
          <div
            aria-hidden="true"
            className={`
              invisible fixed inset-0 bg-[rgba(0,0,0,.4)] opacity-0
              transition-opacity duration-200
              ${
                filterDrawerVisible
                  ? `visible z-side-drawer-backdrop opacity-100`
                  : ""
              }
            `}
            onClick={() => {
              if (!isClosing) {
                handleCloseDrawer();
              }
            }}
          />

          <div
            aria-label="Filters"
            className={`
              fixed top-0 right-0 z-filter-drawer flex h-full max-w-[380px]
              flex-col items-start gap-y-5 bg-default text-left text-inverse
              duration-200 ease-in-out
              ${
                filterDrawerVisible
                  ? `
                    bottom-0 w-full transform-[translateX(0)] overflow-y-auto
                    drop-shadow-2xl
                  `
                  : `w-0 transform-[translateX(100%)] overflow-y-hidden`
              }
              tablet:gap-y-10
            `}
            id="filters"
            ref={filtersRef}
          >
            <div
              className={`
                flex h-full w-full flex-col text-sm
                tablet:text-base
                [@media(min-height:815px)]:pt-12
              `}
            >
              {/* Close button */}
              <button
                aria-label="Close filters"
                className={`
                  absolute top-7 right-4 z-10 flex h-10 w-10 transform-gpu
                  cursor-pointer items-center justify-center rounded-full
                  bg-subtle text-default transition-transform
                  hover:scale-105
                  ${isClosing ? "pointer-events-none" : ""}
                `}
                onClick={() => {
                  if (!isClosing) {
                    handleCloseDrawer();
                    toggleButtonRef.current?.focus();
                  }
                }}
                type="button"
              >
                <svg
                  aria-hidden="true"
                  className={`
                    h-4 w-4 transform-gpu
                    ${isClosing ? "animate-spin-recoil" : ""}
                    ${isOpening ? "animate-spin-wind-up" : ""}
                  `}
                  fill="none"
                  stroke="currentColor"
                  strokeWidth="2"
                  viewBox="0 0 28 28"
                >
                  <path
                    d="M7 21L21 7M7 7l14 14"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                  />
                </svg>
              </button>
              <fieldset
                className={`
                  mt-0 flex h-full grow flex-col gap-5 px-container py-10
                  tablet:gap-8
                  tablet-landscape:grow-0 tablet-landscape:gap-12
                  tablet-landscape:px-12
                `}
              >
                <legend
                  className={`
                    block w-full pt-10 pb-8 font-sans text-lg text-xxs
                    font-semibold tracking-wide text-subtle uppercase
                    shadow-bottom
                  `}
                >
                  Filter
                </legend>
                {filters}
              </fieldset>
              <div
                className={`
                  sticky bottom-0 z-filter-footer mt-auto w-full self-end
                  border-t border-t-default bg-default px-8 py-4 drop-shadow-2xl
                  tablet-landscape:px-12
                `}
              >
                <div className="flex gap-x-4">
                  <button
                    aria-label="Clear all filters"
                    className={`
                      flex items-center justify-center gap-x-4 rounded-sm px-4
                      py-3 font-sans text-xs text-nowrap uppercase
                      ${
                        hasActiveFilters
                          ? "cursor-pointer bg-subtle text-default"
                          : "cursor-not-allowed bg-canvas text-muted opacity-50"
                      }
                    `}
                    disabled={!hasActiveFilters}
                    onClick={() => {
                      if (hasActiveFilters) {
                        onClearFilters?.();
                      }
                    }}
                    type="button"
                  >
                    Clear
                  </button>
                  <button
                    className={`
                      flex flex-1 cursor-pointer items-center justify-center
                      gap-x-4 rounded-sm bg-footer px-4 py-3 font-sans text-xs
                      text-nowrap text-inverse uppercase
                    `}
                    onClick={() => {
                      // Apply pending filters
                      onApplyFilters?.();
                      handleCloseDrawer(false); // Don't reset filters when applying
                      if (typeof document !== "undefined") {
                        document.querySelector("#list")?.scrollIntoView();
                      }
                    }}
                    type="button"
                  >
                    View{" "}
                    {pendingFilteredCount === undefined
                      ? totalCount
                      : pendingFilteredCount}{" "}
                    Results
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

function ListHeader<T extends string>({
  filterDrawerVisible,
  listHeaderButtons,
  onFilterClick,
  sortProps,
  toggleButtonRef,
  totalCount,
}: {
  filterDrawerVisible: boolean;
  listHeaderButtons?: ReactNode;
  onFilterClick: (event: React.MouseEvent) => void;
  sortProps: SortProps<T>;
  toggleButtonRef: React.RefObject<HTMLButtonElement | null>;
  totalCount: number;
}): JSX.Element {
  const { currentSortValue, onSortChange, sortOptions } = sortProps;

  return (
    <div
      className={`
        mx-auto grid max-w-[var(--breakpoint-desktop)]
        grid-cols-[auto_auto_1fr_auto] items-baseline gap-y-7 px-container py-10
        font-sans font-medium tracking-wide text-subtle uppercase
        tablet:grid-cols-[auto_auto_1fr_auto_auto] tablet:gap-x-4
        tablet-landscape:grid-cols-[auto_auto_1fr_minmax(302px,calc(33%_-_192px))_auto]
        desktop:grid-cols-[auto_auto_1fr_calc(33%_-_96px)_auto]
      `}
    >
      <span className={`text-nowrap`}>
        <span className={`font-semibold text-default`}>
          {totalCount.toLocaleString()}
        </span>
        <span className="text-xxs leading-none tracking-wide"> Results</span>
      </span>
      <div className={``}>{listHeaderButtons && listHeaderButtons}</div>
      <div
        className={`
          col-span-full
          tablet:col-span-1 tablet:col-start-4
          desktop:pl-8
        `}
      >
        <label
          className={`
            flex items-baseline gap-x-4 text-xxs font-semibold tracking-wide
            text-subtle
          `}
        >
          Sort{" "}
          <select
            className={`
              flex w-full appearance-none border-none bg-default py-2 pr-4 pl-4
              font-serif text-base font-normal tracking-normal overflow-ellipsis
              text-default shadow-all outline-accent
            `}
            onChange={onSortChange}
            value={currentSortValue}
          >
            {sortOptions}
          </select>
        </label>
      </div>
      <button
        aria-controls="filters"
        aria-expanded={filterDrawerVisible}
        aria-label="Toggle filters"
        className={`
          col-start-4 row-start-1 flex transform-gpu cursor-pointer items-center
          justify-center gap-x-4 rounded-sm bg-canvas px-4 py-2 font-sans
          text-xs font-semibold text-nowrap text-muted uppercase shadow-all
          transition-transform
          hover:scale-110
          tablet:col-start-5 tablet:w-20
        `}
        onClick={onFilterClick}
        ref={toggleButtonRef}
        type="button"
      >
        Filter
      </button>
    </div>
  );
}
