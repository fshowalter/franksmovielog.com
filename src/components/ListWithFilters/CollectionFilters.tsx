import { TextField } from "~/components/Fields/TextField";

import type { CollectionFilterValues } from "./collectionsReducerUtils";

export function CollectionFilters({
  name,
}: {
  name: {
    initialValue: CollectionFilterValues["name"];
    onChange: (value: string) => void;
  };
}): React.JSX.Element {
  return (
    <>
      <TextField
        initialValue={name.initialValue}
        label="Name"
        onInputChange={name.onChange}
        placeholder="Enter all or part of a name"
      />
    </>
  );
}
