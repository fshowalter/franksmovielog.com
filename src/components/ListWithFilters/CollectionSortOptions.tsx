export function CollectionSortOptions({
  options,
}: {
  options: ("name" | "review-count")[];
}): React.JSX.Element {
  return (
    <>
      {options.includes("name") && (
        <>
          <option value="name-asc">Name (A &rarr; Z)</option>
          <option value="name-desc">Name (Z &rarr; A)</option>
        </>
      )}
      {options.includes("review-count") && (
        <>
          <option value="review-count-desc">Review Count (Most First)</option>
          <option value="review-count-asc">Review Count (Fewest First)</option>
        </>
      )}
    </>
  );
}
