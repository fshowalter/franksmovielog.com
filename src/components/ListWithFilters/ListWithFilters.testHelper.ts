import type { UserEvent } from "@testing-library/user-event";

import { act, screen } from "@testing-library/react";
import { vi } from "vitest";

import { DRAWER_CLOSE_ANIMATION_MS } from "./ListWithFilters";

export async function clickClearFilters(user: UserEvent) {
  return user.click(screen.getByRole("button", { name: "Clear all filters" }));
}

export async function clickCloseFilters(user: UserEvent) {
  // Close the drawer with the X button (should reset pending changes)
  await user.click(screen.getByRole("button", { name: "Close filters" }));

  // Wait for drawer close animation
  act(() => {
    vi.advanceTimersByTime(DRAWER_CLOSE_ANIMATION_MS);
  });
}

export async function clickSortOption(user: UserEvent, sortText: string) {
  await user.selectOptions(screen.getByLabelText("Sort"), sortText);
}

export async function clickToggleFilters(user: UserEvent) {
  return user.click(screen.getByRole("button", { name: "Toggle filters" }));
}

export async function clickViewResults(user: UserEvent) {
  await user.click(screen.getByRole("button", { name: /View \d+ Results/ }));
  // Wait for drawer close animation
  act(() => {
    vi.advanceTimersByTime(DRAWER_CLOSE_ANIMATION_MS);
  });
}
