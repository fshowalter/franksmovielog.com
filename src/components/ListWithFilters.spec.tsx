import { act, render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { afterEach, describe, it, vi } from "vitest";

import { ListWithFilters } from "./ListWithFilters";

// Mock scrollIntoView for all tests
if (!Element.prototype.scrollIntoView) {
  Element.prototype.scrollIntoView = vi.fn();
}

// Test props for ListWithFilters
const mockProps = {
  backdrop: <div data-testid="backdrop">Test Backdrop</div>,
  filters: (
    <div data-testid="filters-content">
      <input placeholder="Filter input" type="text" />
      <button type="button">Filter button</button>
    </div>
  ),
  list: <div data-testid="list">Test List Content</div>,
  sortProps: {
    currentSortValue: "title-asc",
    onSortChange: vi.fn(),
    sortOptions: (
      <>
        <option value="title-asc">Title (A → Z)</option>
        <option value="title-desc">Title (Z → A)</option>
      </>
    ),
  },
  totalCount: 100,
};

describe("ListWithFilters", () => {
  it("renders", ({ expect }) => {
    const { asFragment } = render(<ListWithFilters {...mockProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("displays the correct total count", ({ expect }) => {
    expect.hasAssertions();

    render(<ListWithFilters {...mockProps} totalCount={250} />);

    expect(screen.getByText("250")).toBeInTheDocument();
    expect(screen.getByText("Results")).toBeInTheDocument();
  });

  it("can sort items", async ({ expect }) => {
    expect.hasAssertions();

    const onSortChange = vi.fn();
    const props = {
      ...mockProps,
      sortProps: {
        ...mockProps.sortProps,
        onSortChange,
      },
    };

    render(<ListWithFilters {...props} />);

    const sortSelect = screen.getByLabelText("Sort");

    await userEvent.selectOptions(sortSelect, "title-desc");

    expect(onSortChange).toHaveBeenCalled();
  });

  it("renders optional list header buttons", ({ expect }) => {
    expect.hasAssertions();

    const listHeaderButtons = (
      <button data-testid="header-button" type="button">
        Custom Header Button
      </button>
    );

    render(
      <ListWithFilters {...mockProps} listHeaderButtons={listHeaderButtons} />,
    );

    expect(screen.getByTestId("header-button")).toBeInTheDocument();
  });

  it("renders optional sub navigation", ({ expect }) => {
    expect.hasAssertions();

    const dynamicSubNav = <nav data-testid="sub-nav">Sub Navigation</nav>;

    render(<ListWithFilters {...mockProps} dynamicSubNav={dynamicSubNav} />);

    expect(screen.getByTestId("sub-nav")).toBeInTheDocument();
  });

  it("applies custom className", ({ expect }) => {
    expect.hasAssertions();

    const { container } = render(
      <ListWithFilters {...mockProps} className="transform-3d" />,
    );

    const layoutElement = container.querySelector(".transform-3d");
    expect(layoutElement).toBeTruthy();
  });

  describe("filter drawer", () => {
    afterEach(() => {
      // Clean up body classes
      document.body.classList.remove("overflow-hidden");
    });

    it("opens and closes filter drawer", async ({ expect }) => {
      expect.hasAssertions();

      render(<ListWithFilters {...mockProps} />);

      const filterButton = screen.getByRole("button", {
        name: "Toggle filters",
      });
      const filterDrawer = screen.getByLabelText("Filters");

      // Initially closed
      expect(filterButton.getAttribute("aria-expanded")).toBe("false");
      expect(filterDrawer).toMatchSnapshot();

      // Open drawer
      await userEvent.click(filterButton);
      expect(filterButton.getAttribute("aria-expanded")).toBe("true");
      expect(document.body.classList.contains("overflow-hidden")).toBe(true);
      expect(filterDrawer).toMatchSnapshot();

      // Close drawer
      await userEvent.click(filterButton);
      // Wait for animation to complete
      await waitFor(
        () => {
          expect(filterButton.getAttribute("aria-expanded")).toBe("false");
          expect(document.body.classList.contains("overflow-hidden")).toBe(
            false,
          );
        },
        { timeout: 300 },
      );
    });

    it("closes filter drawer with escape key", async ({ expect }) => {
      expect.hasAssertions();

      render(<ListWithFilters {...mockProps} />);

      const filterButton = screen.getByRole("button", {
        name: "Toggle filters",
      });

      // Open drawer
      await userEvent.click(filterButton);
      expect(filterButton.getAttribute("aria-expanded")).toBe("true");

      // Press escape
      await userEvent.keyboard("{Escape}");
      // Wait for animation to complete
      await waitFor(
        () => {
          expect(filterButton.getAttribute("aria-expanded")).toBe("false");
          expect(document.body.classList.contains("overflow-hidden")).toBe(
            false,
          );
        },
        { timeout: 300 },
      );
    });

    it("closes filter drawer when clicking outside (on backdrop)", async ({
      expect,
    }) => {
      expect.hasAssertions();

      const { container } = render(<ListWithFilters {...mockProps} />);
      const filterButton = screen.getByRole("button", {
        name: "Toggle filters",
      });

      // Open drawer
      await userEvent.click(filterButton);
      expect(filterButton.getAttribute("aria-expanded")).toBe("true");
      expect(document.body.classList.contains("overflow-hidden")).toBe(true);

      // Find the backdrop element and click it
      const backdrop = container.querySelector(
        '[aria-hidden="true"].fixed.inset-0',
      );
      expect(backdrop).toBeTruthy();

      if (backdrop) {
        // Simulate clicking the backdrop
        act(() => {
          // Create and dispatch a native click event
          const clickEvent = new MouseEvent("click", {
            bubbles: true,
            cancelable: true,
          });
          backdrop.dispatchEvent(clickEvent);
        });
      }

      // Drawer should be closed - wait for animation
      await waitFor(
        () => {
          expect(filterButton.getAttribute("aria-expanded")).toBe("false");
          expect(document.body.classList.contains("overflow-hidden")).toBe(
            false,
          );
        },
        { timeout: 300 },
      );
    });

    it("closes filter drawer with View Results button", async ({ expect }) => {
      expect.hasAssertions();

      render(<ListWithFilters {...mockProps} />);

      const filterButton = screen.getByRole("button", {
        name: "Toggle filters",
      });

      // Open drawer
      await userEvent.click(filterButton);
      expect(filterButton.getAttribute("aria-expanded")).toBe("true");

      // Click View Results
      const viewResultsButton = screen.getByRole("button", {
        name: /View \d+ Results/,
      });
      await userEvent.click(viewResultsButton);

      // Wait for animation to complete
      await waitFor(
        () => {
          expect(filterButton.getAttribute("aria-expanded")).toBe("false");
          expect(document.body.classList.contains("overflow-hidden")).toBe(
            false,
          );
        },
        { timeout: 300 },
      );
    });
  });
});
