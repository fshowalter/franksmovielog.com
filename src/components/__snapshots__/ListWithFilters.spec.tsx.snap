// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListWithFilters > filter drawer > opens and closes filter drawer 1`] = `
<div
  aria-label="Filters"
  class="
              fixed top-0 right-0 z-filter-drawer flex h-full max-w-[380px]
              flex-col items-start gap-y-5 bg-default text-left text-inverse
              duration-200 ease-in-out
              w-0 transform-[translateX(100%)] overflow-y-hidden
              tablet:gap-y-10
            "
  id="filters"
>
  <div
    class="
                flex h-full w-full flex-col text-sm
                tablet:text-base
                [@media(min-height:815px)]:pt-12
              "
  >
    <button
      aria-label="Close filters"
      class="
                  absolute top-7 right-4 z-10 flex h-10 w-10 transform-gpu
                  cursor-pointer items-center justify-center rounded-full
                  bg-canvas text-default drop-shadow-md transition-transform
                  hover:scale-105
                  
                "
      type="button"
    >
      <svg
        aria-hidden="true"
        class="
                    h-4 w-4 transform-gpu
                    
                    
                  "
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 28 28"
      >
        <path
          d="M7 21L21 7M7 7l14 14"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <fieldset
      class="
                  mt-0 flex h-full grow flex-col gap-5 px-container py-10
                  tablet:gap-8
                  tablet-landscape:grow-0 tablet-landscape:gap-12
                  tablet-landscape:px-12
                "
    >
      <legend
        class="
                    block w-full pt-10 pb-8 font-sans text-lg text-xxs
                    font-semibold tracking-wide text-subtle uppercase
                    shadow-bottom
                  "
      >
        Filter
      </legend>
      <div
        data-testid="filters-content"
      >
        <input
          placeholder="Filter input"
          type="text"
        />
        <button
          type="button"
        >
          Filter button
        </button>
      </div>
    </fieldset>
    <div
      class="
                  sticky bottom-0 z-filter-footer mt-auto w-full self-end
                  border-t border-t-default bg-default px-8 py-4 drop-shadow-2xl
                  tablet-landscape:px-12
                "
    >
      <div
        class="flex gap-x-4"
      >
        <button
          aria-label="Clear all filters"
          class="
                      flex items-center justify-center gap-x-4 rounded-sm
                      bg-canvas px-4 py-3 font-sans text-xs text-nowrap
                      uppercase
                      cursor-not-allowed text-muted opacity-50
                    "
          disabled=""
          type="button"
        >
          Clear
        </button>
        <button
          class="
                      flex flex-1 cursor-pointer items-center justify-center
                      gap-x-4 rounded-sm bg-footer px-4 py-3 font-sans text-xs
                      text-nowrap text-inverse uppercase
                    "
          type="button"
        >
          View
           
          100
           
          Results
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListWithFilters > filter drawer > opens and closes filter drawer 2`] = `
<div
  aria-label="Filters"
  class="
              fixed top-0 right-0 z-filter-drawer flex h-full max-w-[380px]
              flex-col items-start gap-y-5 bg-default text-left text-inverse
              duration-200 ease-in-out
              
                    bottom-0 w-full transform-[translateX(0)] overflow-y-auto
                    drop-shadow-2xl
                  
              tablet:gap-y-10
            "
  id="filters"
>
  <div
    class="
                flex h-full w-full flex-col text-sm
                tablet:text-base
                [@media(min-height:815px)]:pt-12
              "
  >
    <button
      aria-label="Close filters"
      class="
                  absolute top-7 right-4 z-10 flex h-10 w-10 transform-gpu
                  cursor-pointer items-center justify-center rounded-full
                  bg-canvas text-default drop-shadow-md transition-transform
                  hover:scale-105
                  
                "
      type="button"
    >
      <svg
        aria-hidden="true"
        class="
                    h-4 w-4 transform-gpu
                    
                    animate-spin-wind-up
                  "
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 28 28"
      >
        <path
          d="M7 21L21 7M7 7l14 14"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <fieldset
      class="
                  mt-0 flex h-full grow flex-col gap-5 px-container py-10
                  tablet:gap-8
                  tablet-landscape:grow-0 tablet-landscape:gap-12
                  tablet-landscape:px-12
                "
    >
      <legend
        class="
                    block w-full pt-10 pb-8 font-sans text-lg text-xxs
                    font-semibold tracking-wide text-subtle uppercase
                    shadow-bottom
                  "
      >
        Filter
      </legend>
      <div
        data-testid="filters-content"
      >
        <input
          placeholder="Filter input"
          type="text"
        />
        <button
          type="button"
        >
          Filter button
        </button>
      </div>
    </fieldset>
    <div
      class="
                  sticky bottom-0 z-filter-footer mt-auto w-full self-end
                  border-t border-t-default bg-default px-8 py-4 drop-shadow-2xl
                  tablet-landscape:px-12
                "
    >
      <div
        class="flex gap-x-4"
      >
        <button
          aria-label="Clear all filters"
          class="
                      flex items-center justify-center gap-x-4 rounded-sm
                      bg-canvas px-4 py-3 font-sans text-xs text-nowrap
                      uppercase
                      cursor-not-allowed text-muted opacity-50
                    "
          disabled=""
          type="button"
        >
          Clear
        </button>
        <button
          class="
                      flex flex-1 cursor-pointer items-center justify-center
                      gap-x-4 rounded-sm bg-footer px-4 py-3 font-sans text-xs
                      text-nowrap text-inverse uppercase
                    "
          type="button"
        >
          View
           
          100
           
          Results
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListWithFilters > renders 1`] = `
<DocumentFragment>
  <div
    class="
        
      "
  >
    <div
      class="group/list-with-filters mx-auto bg-subtle"
    >
      <div
        class="
            sticky top-[calc(0px_+_var(--scroll-offset,0px))] z-sticky
            scroll-mt-[calc(0px_+_var(--scroll-offset,0px))] border-b
            border-default bg-default text-xs
            tablet:col-span-full
          "
      >
        <div
          class="
        mx-auto grid max-w-[var(--breakpoint-desktop)]
        grid-cols-[auto_auto_1fr_auto] items-baseline gap-y-7 px-container py-10
        font-sans font-medium tracking-wide text-subtle uppercase
        tablet:grid-cols-[auto_auto_1fr_auto_auto] tablet:gap-x-4
        tablet-landscape:grid-cols-[auto_auto_1fr_minmax(302px,calc(33%_-_192px))_auto]
        desktop:grid-cols-[auto_auto_1fr_calc(33%_-_96px)_auto]
      "
        >
          <span
            class="text-nowrap"
          >
            <span
              class="font-semibold text-default"
            >
              100
            </span>
            <span
              class="text-xxs leading-none tracking-wide"
            >
               Results
            </span>
          </span>
          <div
            class=""
          />
          <div
            class="
          col-span-full
          tablet:col-span-1 tablet:col-start-4
          desktop:pl-8
        "
          >
            <label
              class="
            flex items-baseline gap-x-4 text-xxs font-semibold tracking-wide
            text-subtle
          "
            >
              Sort 
              <select
                class="
              flex w-full appearance-none border-none bg-default py-2 pr-4 pl-4
              font-serif text-base font-normal tracking-normal overflow-ellipsis
              text-default shadow-all outline-accent
            "
              >
                <option
                  value="title-asc"
                >
                  Title (A → Z)
                </option>
                <option
                  value="title-desc"
                >
                  Title (Z → A)
                </option>
              </select>
            </label>
          </div>
          <button
            aria-controls="filters"
            aria-expanded="false"
            aria-label="Toggle filters"
            class="
          col-start-4 row-start-1 flex transform-gpu cursor-pointer items-center
          justify-center gap-x-4 rounded-sm bg-canvas px-4 py-2 font-sans
          text-xs font-semibold text-nowrap text-muted uppercase shadow-all
          transition-transform
          hover:scale-110
          tablet:col-start-5 tablet:w-20
        "
            type="button"
          >
            Filter
          </button>
        </div>
      </div>
      <div
        class="
            mx-auto max-w-[var(--breakpoint-desktop)]
            tablet:px-container
          "
      >
        <div
          class="
              mx-auto max-w-[var(--breakpoint-desktop)] grow
              scroll-mt-[calc(var(--list-scroll-offset)_+_var(--scroll-offset,0px))]
              pb-10
              [--list-scroll-offset:181px]
              tablet:[--list-scroll-offset:121px]
            "
          id="list"
        >
          <div
            data-testid="list"
          >
            Test List Content
          </div>
        </div>
        <div
          aria-hidden="true"
          class="
              invisible fixed inset-0 bg-[rgba(0,0,0,.4)] opacity-0
              transition-opacity duration-200
              
            "
        />
        <div
          aria-label="Filters"
          class="
              fixed top-0 right-0 z-filter-drawer flex h-full max-w-[380px]
              flex-col items-start gap-y-5 bg-default text-left text-inverse
              duration-200 ease-in-out
              w-0 transform-[translateX(100%)] overflow-y-hidden
              tablet:gap-y-10
            "
          id="filters"
        >
          <div
            class="
                flex h-full w-full flex-col text-sm
                tablet:text-base
                [@media(min-height:815px)]:pt-12
              "
          >
            <button
              aria-label="Close filters"
              class="
                  absolute top-7 right-4 z-10 flex h-10 w-10 transform-gpu
                  cursor-pointer items-center justify-center rounded-full
                  bg-canvas text-default drop-shadow-md transition-transform
                  hover:scale-105
                  
                "
              type="button"
            >
              <svg
                aria-hidden="true"
                class="
                    h-4 w-4 transform-gpu
                    
                    
                  "
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 28 28"
              >
                <path
                  d="M7 21L21 7M7 7l14 14"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
            <fieldset
              class="
                  mt-0 flex h-full grow flex-col gap-5 px-container py-10
                  tablet:gap-8
                  tablet-landscape:grow-0 tablet-landscape:gap-12
                  tablet-landscape:px-12
                "
            >
              <legend
                class="
                    block w-full pt-10 pb-8 font-sans text-lg text-xxs
                    font-semibold tracking-wide text-subtle uppercase
                    shadow-bottom
                  "
              >
                Filter
              </legend>
              <div
                data-testid="filters-content"
              >
                <input
                  placeholder="Filter input"
                  type="text"
                />
                <button
                  type="button"
                >
                  Filter button
                </button>
              </div>
            </fieldset>
            <div
              class="
                  sticky bottom-0 z-filter-footer mt-auto w-full self-end
                  border-t border-t-default bg-default px-8 py-4 drop-shadow-2xl
                  tablet-landscape:px-12
                "
            >
              <div
                class="flex gap-x-4"
              >
                <button
                  aria-label="Clear all filters"
                  class="
                      flex items-center justify-center gap-x-4 rounded-sm
                      bg-canvas px-4 py-3 font-sans text-xs text-nowrap
                      uppercase
                      cursor-not-allowed text-muted opacity-50
                    "
                  disabled=""
                  type="button"
                >
                  Clear
                </button>
                <button
                  class="
                      flex flex-1 cursor-pointer items-center justify-center
                      gap-x-4 rounded-sm bg-footer px-4 py-3 font-sans text-xs
                      text-nowrap text-inverse uppercase
                    "
                  type="button"
                >
                  View 100 Results
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
